packages/client/lib/sync/fullsync.ts

```diff
@@ -204,11 +213,57 @@ export class FullSynchronizer extends Synchronizer {
    * Processes `NEW_BLOCK` announcement from a peer and inserts into local chain if child of chain tip
    * @param blockData `NEW_BLOCK` received from peer
    */
-  async handleNewBlock(block: Block) {
-    const chainTip = (await this.chain.getLatestHeader()).hash()
+  async handleNewBlock(block: Block, peer?: Peer) {
+    if (peer) {
+      // Don't send NEW_BLOCK announcement to peer that sent original new block message
+      const knownBlocks = this.newBlocksKnownByPeer.get(peer.id) ?? []
+      knownBlocks.push({ hash: block.hash(), added: Date.now() })
+      this.newBlocksKnownByPeer.set(peer.id, knownBlocks)
+    }
+
+    const chainTip = await this.chain.getLatestBlock()
+
     // If block parent is current chain tip, insert block into chain
-    if (chainTip.toString('hex') === block.header.parentHash.toString('hex')) {
+    if (chainTip.header.hash().equals(block.header.parentHash)) {
+      if (!block.validateDifficulty(chainTip)) return
+
+      // Send NEW_BLOCK to square root of total number of peers in pool
+      // https://github.com/ethereum/devp2p/blob/master/caps/eth.md#block-propagation
+      const numPeersToShareWith = Math.floor(Math.sqrt(this.pool.peers.length))
+
+      let x = 0
+      while (x < numPeersToShareWith) {
+        const currentPeer = this.pool.peers[x]
+        const knownBlocks = this.newBlocksKnownByPeer.get(currentPeer.id) ?? []
+        if (
+          knownBlocks.find(
+            (handledObject) => handledObject.hash && handledObject.hash.equals(block.hash())
+          )
+        ) {
+          // If peer has already been sent this block, skip peer
+          x++
+          continue
+        }
+
+        knownBlocks.push({ hash: block.hash(), added: Date.now() })
+        this.newBlocksKnownByPeer.set(currentPeer.id, knownBlocks)
+        currentPeer.eth?.send('NewBlock', [block, this.chain.blocks.td])
+
+        x++
+      }
```

Comment:
Same here, this can be extracted as an [addToKnownByPeer()](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/client/lib/sync/txpool.ts#L226) function analogue to the tx pool.

These extractions have the advantage that this outer `handleNewBlock()` function becomes better readable and shorter and it becomes furthermore also possible to do more granular testing.

(sorry, don't want per-se impose my introduced structures on you, just think in this case it does just make sense to align since the functionality is so similar and the client code will get easier to read if we align on things like this).